packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -860,84 +1029,81 @@ class TextPainter {
-    _paragraph!.layout(ui.ParagraphConstraints(width: maxWidth));
-    if (minWidth != maxWidth) {
-      double newWidth;
-      switch (textWidthBasis) {
-        case TextWidthBasis.longestLine:
-          // The parent widget expects the paragraph to be exactly
-          // `TextPainter.width` wide, if that value satisfies the constraints
-          // it gave to the TextPainter. So when `textWidthBasis` is longestLine,
-          // the paragraph's width needs to be as close to the width of its
-          // longest line as possible.
-          newWidth = _applyFloatingPointHack(_paragraph!.longestLine);
-          break;
-        case TextWidthBasis.parent:
-          newWidth = maxIntrinsicWidth;
-          break;
-      }
-      newWidth = clampDouble(newWidth, minWidth, maxWidth);
-      if (newWidth != _applyFloatingPointHack(_paragraph!.width)) {
-        _paragraph!.layout(ui.ParagraphConstraints(width: newWidth));
-      }
-    }
+    return builder.build();
   }
 
   /// Computes the visual position of the glyphs for painting the text.
   ///
   /// The text will layout with a width that's as close to its max intrinsic
-  /// width as possible while still being greater than or equal to `minWidth` and
-  /// less than or equal to `maxWidth`.
+  /// width (or its longest line, if [textWidthBasis] is set to
+  /// [TextWidthBasis.parent]) as possible while still being greater than or
+  /// equal to `minWidth` and less than or equal to `maxWidth`.
   ///
   /// The [text] and [textDirection] properties must be non-null before this is
   /// called.
   void layout({ double minWidth = 0.0, double maxWidth = double.infinity }) {
-    assert(text != null, 'TextPainter.text must be set to a non-null value before using the TextPainter.');
-    assert(textDirection != null, 'TextPainter.textDirection must be set to a non-null value before using the TextPainter.');
-    // Return early if the current layout information is not outdated, even if
-    // _needsPaint is true (in which case _paragraph will be rebuilt in paint).
-    if (_paragraph != null && minWidth == _lastMinWidth && maxWidth == _lastMaxWidth) {
-      return;
+    assert(!maxWidth.isNaN);
+    assert(!minWidth.isNaN);
+    final InlineSpan? text = this.text;
+    if (text == null) {
+      throw StateError('TextPainter.text must be set to a non-null value before using the TextPainter.');
+    }
+    final TextDirection? textDirection = this.textDirection;
+    if (textDirection == null) {
+      throw StateError('TextPainter.textDirection must be set to a non-null value before using the TextPainter.');
     }
+    final _TextPainterLayoutCache? cachedLayout = _layoutCache;
 
-    if (_rebuildParagraphForPaint || _paragraph == null) {
-      _createParagraph();
+    assert(() {
+      _debugNeedsRelayout = false;
+      return true;
+    }());
+    if (cachedLayout != null && cachedLayout._resizeToFit(minWidth, maxWidth, textWidthBasis)) {
+      return;
     }
```

Comment:
For my own understanding: This is the only place where we benefit from this new cache, right? Also that, plus the avoided layout if the condition in line 1096 is false are the only functional changes that this PR introduces. Everything else is staying pretty much the same?